### PR TITLE
[Docs only] Rearranging structure

### PR DESCRIPTION
**Pull request by wild-endeavor**
Not really adding any content, just playing around with the structure.  Moving as much of the generation (autosummary) directives and prose alongside the code as possible.  This makes it less likely that we expose a new import to the user without documenting it.

### CODE REVIEW

File: flytekit/extend/__init__.py [hunk 1]
```diff
@@ -1,4 +1,34 @@
...
+=====================
+
+.. currentmodule:: flytekit.extend
+
+This package contains things that are useful when extending Flytekit.
+
+.. autosummary::
+   :toctree: generated/
+
+   get_serializable - fdjsalkfsd
```
Comment (kumare3) [inline flytekit/extend/__init__.py:13]: ?

File: flytekit/extend/__init__.py [hunk 2]
```diff
@@ -1,4 +1,34 @@
...
+   PythonTask
+   ExecutionState
+   Image
+   ImageConfig
+   SerializationSettings
+   Interface
+   Promise
+   TaskPlugins
+   DictTransformer
+   T
```
Comment (kumare3) [inline flytekit/extend/__init__.py:26]: ?

File: README.md [hunk 3]
```diff
@@ -51,55 +51,3 @@ def my_workflow(x: int, y: int) -> int:
 Refer to [plugins/README.md](plugins/README.md) for a list of available
 plugins. There may be plugins outside of this list, but this list is maintained
 by the core maintainers.
```
Comment (kumare3) [inline README.md:53]: where did all the contrib docs go? Did we move it to contributing?

File: flytekit/extend/__init__.py (hunk 2 above)
Comment (wild-endeavor) [reply flytekit/extend/__init__.py:26]: I have no idea what this is.  But one of the plugins uses it!  I've yet to figure it out.

File: README.md (hunk 3 above)
Comment (wild-endeavor) [reply README.md:53]: Not yet, i copied it into a text file locally.  I was gonna work in it next.

File: flytekit/extend/__init__.py (hunk 1 above)
Comment (wild-endeavor) [reply flytekit/extend/__init__.py:13]: does this not make sense to you?

File: docs/source/index.rst [hunk 4]
```diff
@@ -3,44 +3,55 @@
...
-   :caption: Flytekit Learn by Example
+   :caption: APIs
 
-   Flytekit Learn by Example <http://flytecookbook.readthedocs.io/>
-   Flyte Documentation <https://lyft.github.io/flyte/>
+   flytekit
+   testing
+   extend
+
+This section documents the various Python imports users will need to work with to build Flyte workflows, extend Flytekit, etc.
```
Comment (katrogan) [inline docs/source/index.rst:22]: nit: thoughts on moving this above the TOC? took me a moment to decipher which TOC it was referring to

Comment (wild-endeavor) [reply docs/source/index.rst:22]: Actually i did it both ways.  I think in rendered form, it looks better here.

File: flytekit/core/dynamic_workflow_task.py [hunk 5]
```diff
@@ -1,3 +1,23 @@
...
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+.. autoclass:: flytekit.core.dynamic_workflow_task.DynamicWorkflowTask
+
+
+Caveats when using a dynamic workflow
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Because of the dynamic nature of the workflow generated, it can easily be abused. Keep in mind that the workflow
+that's compiled out of the decorated function needs to be processed like any other workflow. It's rare to see a
+manually written workflow that has 5000 nodes for instance, but you can easily get there with a loop. Please keep,
```
Comment (katrogan) [inline flytekit/core/dynamic_workflow_task.py:17]: should we add a TODO here to update w/ map tasks when available if you are doing some massively parallel thing

**Review by katrogan — APPROVED** [review]

File: flytekit/core/dynamic_workflow_task.py (hunk 5 above)
Comment (wild-endeavor) [reply flytekit/core/dynamic_workflow_task.py:17]: but this will be done in like three days right?  we'll remember.

**Review by kumare3 — APPROVED** [review]

**Pull request merged**